### ISSUE TITLE
WakaTime Error : SSL behind corporate proxy

### ISSUE THREAD

**Issue by BFlavien** [wakatime/vscode-wakatime#13]
Hello,

I'm working behind a corporate proxy and can't manage to make it work.

I installed python, and wakatime-cli.
Everything runs well when using with SublimeText plugin, but I can't make it work with VSCode.

Here are the logs from VSCode Developers Tool :

```
extensionHost.ts:282 [Extension Host] Uncaught Exception:  Error: write EPROTO 0:error:14077419:SSL routines:SSL23_GET_SERVER_HELLO:tlsv1 alert access denied:c:\jenkins\workspace\electron-win-ia32\vendor\node\deps\openssl\openssl\ssl\s23_clnt.c:769:
shell.ts:437 write EPROTO 0:error:14077419:SSL routines:SSL23_GET_SERVER_HELLO:tlsv1 alert access denied:c:\jenkins\workspace\electron-win-ia32\vendor\node\deps\openssl\openssl\ssl\s23_clnt.c:769:
: Error: write EPROTO 0:error:14077419:SSL routines:SSL23_GET_SERVER_HELLO:tlsv1 alert access denied:c:\jenkins\workspace\electron-win-ia32\vendor\node\deps\openssl\openssl\ssl\s23_clnt.c:769:
    at exports._errnoException (util.js:1026:11)
    at WriteWrap.afterWrite (net.js:795:14)
extensionHost.ts:282 [Extension Host] Error: write EPROTO 0:error:14077419:SSL routines:SSL23_GET_SERVER_HELLO:tlsv1 alert access denied:c:\jenkins\workspace\electron-win-ia32\vendor\node\deps\openssl\openssl\ssl\s23_clnt.c:769:
    at exports._errnoException (util.js:1026:11)
    at WriteWrap.afterWrite (net.js:795:14)
shell.ts:437 socket hang up: Error: socket hang up
    at createHangUpError (_http_client.js:252:15)
    at TLSSocket.socketCloseListener (_http_client.js:284:23)
    at emitOne (events.js:101:20)
    at TLSSocket.emit (events.js:188:7)
    at TCP._handle.close [as _onclose] (net.js:493:12)
extensionHost.ts:282 [Extension Host] Uncaught Exception:  Error: socket hang up
extensionHost.ts:282 [Extension Host] Error: socket hang up
    at createHangUpError (_http_client.js:252:15)
    at TLSSocket.socketCloseListener (_http_client.js:284:23)
    at emitOne (events.js:101:20)
    at TLSSocket.emit (events.js:188:7)
    at TCP._handle.close [as _onclose] (net.js:493:12)
```

My wakatime.log got logs from Sublime, but none from VSCode.
[wakatime.log.txt](https://github.com/wakatime/vscode-wakatime/files/771524/wakatime.log.txt)

Thank you :D

**Comment by alanhamlett** [comment]
Try updating your wakatime plugin to v1.0.8, which will use the proxy setting from your `~/.wakatime.cfg` file when downloading it's dependencies. Let me know if this new version fixes the problem!

**Comment by BFlavien** [comment]
Thanks for the (very) quick answer :)

Updated plugin to 1.0.8, and core to 6.2.2 through pip. I had to do 

`set HTTPS_PROXY=my.corporate.proxy`

and 

`export https_proxy=my.corporate.proxy`

to be able to upgrade it.

After reloading VSCode, version was 1.0.8, and didn't prompt me my API key. Unfortunately i still got errors in console : 

`extensionHost.ts:282 [Extension Host] Uncaught Exception:  Error: write EPROTO 0:error:14077419:SSL routines:SSL23_GET_SERVER_HELLO:tlsv1 alert access denied:c:\jenkins\workspace\electron-win-ia32\vendor\node\deps\openssl\openssl\ssl\s23_clnt.c:769:`

```
shell.ts:437 write EPROTO 0:error:14077419:SSL routines:SSL23_GET_SERVER_HELLO:tlsv1 alert access denied:c:\jenkins\workspace\electron-win-ia32\vendor\node\deps\openssl\openssl\ssl\s23_clnt.c:769:
: Error: write EPROTO 0:error:14077419:SSL routines:SSL23_GET_SERVER_HELLO:tlsv1 alert access denied:c:\jenkins\workspace\electron-win-ia32\vendor\node\deps\openssl\openssl\ssl\s23_clnt.c:769:

    at exports._errnoException (util.js:1026:11)
    at WriteWrap.afterWrite (net.js:795:14)
```

```
extensionHost.ts:282 [Extension Host] Error: write EPROTO 0:error:14077419:SSL routines:SSL23_GET_SERVER_HELLO:tlsv1 alert access denied:c:\jenkins\workspace\electron-win-ia32\vendor\node\deps\openssl\openssl\ssl\s23_clnt.c:769:

    at exports._errnoException (util.js:1026:11)
    at WriteWrap.afterWrite (net.js:795:14)
```

```
shell.ts:437 socket hang up: Error: socket hang up
    at createHangUpError (_http_client.js:252:15)
    at TLSSocket.socketCloseListener (_http_client.js:284:23)
    at emitOne (events.js:101:20)
    at TLSSocket.emit (events.js:188:7)
    at TCP._handle.close [as _onclose] (net.js:493:12)
```

`extensionHost.ts:282 [Extension Host] Uncaught Exception:  Error: socket hang up`

```
extensionHost.ts:282 [Extension Host] Error: socket hang up
    at createHangUpError (_http_client.js:252:15)
    at TLSSocket.socketCloseListener (_http_client.js:284:23)
    at emitOne (events.js:101:20)
    at TLSSocket.emit (events.js:188:7)
    at TCP._handle.close [as _onclose] (net.js:493:12)
```

`extensionHost.ts:282 [Extension Host] pythonw: can't open file 'C:\Users\myuser\.vscode\extensions\WakaTime.vscode-wakatime-1.0.8\out\wakatime-master\wakatime\cli.py': [Errno 2] No such file or directory`

This error is the reason I tried to install the core separately.
In `C:\Users\myuser\.vscode\extensions\WakaTime.vscode-wakatime-1.0.8\out\` I got a wakatime-master.zip but no folders. It feels like it can't be unzipped.

When I try to do it manually it says "Cannot Complete the Compressed (zipped) Folders Extraction Wizard. Access to the compressed (zipped) folder is denied. Before you can extract files, you must change the permissions for this compressed (zipped) folder. To close this wizard, click Finish".

```
[Extension Host] Error: Command failed: pythonw C:\Users\myuser\.vscode\extensions\WakaTime.vscode-wakatime-1.0.8\out\wakatime-master\wakatime\cli.py --file c:\Users\myuser\Repositories\uoc-addon-nfvd\src\server\addons\plugins\nfvd\package-manager.js --plugin vscode/1.9.1 vscode-wakatime/1.0.8 --alternate-project src
pythonw: can't open file 'C:\Users\myuser\.vscode\extensions\WakaTime.vscode-wakatime-1.0.8\out\wakatime-master\wakatime\cli.py': [Errno 2] No such file or directory

    at ChildProcess.exithandler (child_process.js:218:12)
    at emitTwo (events.js:111:20)
    at ChildProcess.emit (events.js:191:7)
    at maybeClose (internal/child_process.js:877:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:226:5)
```

`extensionHost.ts:282 [Extension Host] Unknown Error (2); Check your ~/.wakatime.log file for more details.`

Hope this helps ! thanks again :)

**Comment by alanhamlett** [comment]
> I downloaded wakatime.master.zip from official repo, and unzipped it myself.

That's good, also check your ~/.wakatime.cfg file and make sure you have a `debug = true` line, and a `proxy = my.corporate.proxy` line.

Can you paste the whole contents of your Developer Tools Console, which should give us more information about what is happening when the error occurs.

**Comment by alanhamlett** [comment]
You can now edit the debug and proxy setting by pressing `CMD+SHIFT+p` then typing `wakatime`.

**Comment by BFlavien** [comment]
Well, that's good, but I wonder if it's supposed to download and unzip wakatime-core since I installed it separately as a standalone.

My ~.wakatime.cfg already had debug=true : 

```
[settings]
debug = true
api_key = myApiKey
proxy = my.http.corporate.proxy
timeout = 30

[projectmap]
    ~\Repositories\pro.project1\ = project1
    ~\Repositories\pro.project2\ = project2
```

I upgraded the extension to 1.1.1 through VSCode marketplace.

This is the full logs when I launch VS Code
[logs-run.txt](https://github.com/wakatime/vscode-wakatime/files/776513/logs-run.txt)

When I click on a file I got the following logs :
[logs-newFile.txt](https://github.com/wakatime/vscode-wakatime/files/776514/logs-newFile.txt)

After upgrading 1.1.1, I had to manually get back to your github wakatime core, download last release (6.2.2), paste it and unzip to C:\Users\myUser\.vscode\extensions\WakaTime.vscode-wakatime-1.1.1\out\wakatime-master.

After doing so, it's working well.

I still got two questions: 

- I installed wakatime-core separately. Is it not supposed to use it instead of the one downloaded with the plugin?
- Do I have to download, copy, and unzip your core in the extension folder everytime you update either the core or the vscode extension?
[wakatime.log.txt](https://github.com/wakatime/vscode-wakatime/files/776522/wakatime.log.txt)

**Comment by alanhamlett** [comment]
The Vscode console logs mean wakatime-cli is missing:
`pythonw: can't open file 'C:\Users\myuser\.vscode\extensions\WakaTime.vscode-wakatime-1.1.1\out\wakatime-master\wakatime\cli.py': [Errno 2] No such file or directory`

Which is why it starts working once you manually download and extract wakatime-cli.
When the proxy is working, you should see debug logs like this in your VScode Developer Console:
```
[Extension Host] [WakaTime] [DEBUG] Initializing WakaTime v1.1.1
[Extension Host] [WakaTime] [DEBUG] Current wakatime-core version is 6.2.2
[Extension Host] [WakaTime] [DEBUG] Checking for updates to wakatime-core...
[Extension Host] [WakaTime] [DEBUG] wakatime-core is up to date.
```

You would have to manually extract wakatime-cli, unless we can get the proxy working.

Which type of proxy is yours, HTTPS/SOCKS5/NTLM?

Can Vscode download extensions through the proxy? If so, you set the proxy setting in vscode settings? [[1]]

1 https://github.com/Microsoft/vscode/issues/69

[1]: https://github.com/Microsoft/vscode/issues/69

**Comment by BFlavien** [comment]
I don't know much about proxy, but I'd say it's a HTTP one.

I set it up in sublime (that worked straight out of the box) and in vscode ("http.proxy": "http://web-proxy.atl.something.com:8080"),
I got no issue downloading updates, extensions and extensions update.

**Comment by k33k00** [comment]
I am having the same issue. Mine is NTLM, most of the time I get around it by pointing my proxy to Fiddler.  This is how I am able to download extensions from the marketplace. Also how I can access git repos on github.

However i'm getting this in the debug log

```
[Extension Host] [WakaTime] [DEBUG] Initializing WakaTime v1.1.12
```
```
[Extension Host] [WakaTime] [DEBUG] Current wakatime-core version is 8.0.3
```
```
[Extension Host] [WakaTime] [DEBUG] Checking for updates to wakatime-core...
```
```
[Extension Host] [WakaTime] [DEBUG] Unable to find latest wakatime-core version from GitHub.
```
```
[Extension Host] [WakaTime] [DEBUG] Downloading wakatime-core...
```
```
[Extension Host] Uncaught Exception:  Error: connect ECONNREFUSED 192.30.253.112:443
```
```
connect ECONNREFUSED 192.30.253.112:443: Error: connect ECONNREFUSED 192.30.253.112:443
    at Object.exports._errnoException (util.js:1022:11)
    at exports._exceptionWithHostPort (util.js:1045:20)
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1090:14)
```
```
socket hang up: Error: socket hang up
    at createHangUpError (_http_client.js:253:15)
    at TLSSocket.socketCloseListener (_http_client.js:285:23)
    at emitOne (events.js:101:20)
    at TLSSocket.emit (events.js:188:7)
    at _handle.close (net.js:501:12)
    at TCP.done [as _onclose] (_tls_wrap.js:325:7)
```
```
[Extension Host] Error: connect ECONNREFUSED 192.30.253.112:443
    at Object.exports._errnoException (util.js:1022:11)
    at exports._exceptionWithHostPort (util.js:1045:20)
    at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1090:14)
```
```
[Extension Host] Uncaught Exception:  Error: socket hang up
```
```
[Extension Host] Error: socket hang up
    at createHangUpError (_http_client.js:253:15)
    at TLSSocket.socketCloseListener (_http_client.js:285:23)
    at emitOne (events.js:101:20)
    at TLSSocket.emit (events.js:188:7)
    at _handle.close (net.js:501:12)
    at TCP.done [as _onclose] (_tls_wrap.js:325:7)
```
```
[Extension Host] Error: ENOENT: no such file or directory, scandir 'C:\Users\k33k00\AppData\Local\GitHub'
```
I am also unable to install manually as wakatime can't seem to communicate with github to check the version. So it just wipes the zip and retries the download

Also here is my config file
```
[settings]
api_key = ###
debug = true
proxy = http://127.0.0.1:8888
no_ssl_verify = false
```

**Comment by k33k00** [comment]
I've almost got it sorted I think :)

**Comment by alanhamlett** [comment]
Fixed with #22.
